File: Data Structures & Algorithms/balanced-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isBalanced(TreeNode* root) {
        if (root == nullptr) return true;
        int dep = depth(root);
        if (dep==-1) return false;
        return true;
    }
    int depth(TreeNode* root) {
        if (root==nullptr) return 0;
        int lDep = depth(root->left);
        int rDep = depth(root->right);
        if (lDep == -1 || rDep == -1 || 
            lDep - rDep > 1 || lDep - rDep < -1) return -1;
        return 1 + max(lDep,rDep);
    }
};

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        # Swap such that p.val <= q.val
        if p.val > q.val:
            temp = p
            p = q
            q = temp
        
        def traverse(node: TreeNode) -> TreeNode:
            if p.val <= node.val and node.val <= q.val:
                return node
            elif p.val <= node.val and q.val <= node.val:
                return traverse(node.left)
            elif node.val <= p.val and node.val <= q.val:
                return traverse(node.right)

        return traverse(root)